src/dmd/clone.d

```diff
@@ -978,6 +982,53 @@ extern (C++) FuncDeclaration buildDtor(AggregateDeclaration ad, Scope* sc)
     return xdtor;
 }
 
+/**
+ * build a shim function around the compound dtor that accepts an argument
+ *  that is used to implement the deleting C++ destructor
+ *
+ * Params:
+ *  ad = the aggregate that contains the destructor to wrap
+ *  dtor = the destructor to wrap
+ *  sc = the scope in which to analyze the new function
+ *
+ * Returns:
+ *  the shim destructor, semantically analyzed and added to the class as a member
+ */
+private DtorDeclaration buildWindowsCppDtor(AggregateDeclaration ad, DtorDeclaration dtor, Scope* sc)
+{
+    if (!dtor)
+        return null;
+
+    Parameter param = new Parameter(STC.undefined_, Type.tuns32, Identifier.idPool("del"), new IntegerExp(dtor.loc, 0, Type.tuns32));
+    Parameters* params = new Parameters;
+    (*params).push(param);
+    auto ftype = new TypeFunction(params, Type.tvoidptr, false, LINK.cpp, dtor.storage_class);
+    auto func = new DtorDeclaration(dtor.loc, dtor.loc, dtor.storage_class, Id.cppdtor);
+    func.type = ftype;
+    if (dtor.fbody)
+    {
```

Comment:
A comment here showing what D code is being generated would be nice, as it is clumsy to try and piece it together in one's mind from the AST generating expressions.